keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -0,0 +1,1354 @@
+managed by a driver according to a specific "structured model". In the example
+seen below, the structured model in use is called `namedResourcesWithAttributes`:
+
+```yaml
+kind: NodeResourceSlice
+apiVersion: resource.k8s.io/v1alpha2
+...
+spec:
+  nodeName: worker-1
+  driverName: cards.dra.example.com
+  namedResourcesWithAttributes:
+    ...
+```
+
+Such a model could be created to represent resources in a manner similar to the
+opaque strings passed over the tradition device plugin API to the kubelet. The
+one addition being that each named resource can have a set of arbitrary
+attributes attached to it.
+
+If a driver wanted to use a different structured model to represent its resources,
+a new structured model would need to be defined inside Kuberenetes, and a field
+would need to be added to this struct at the same level as
+`namedResourcesWithAttributes`. Driver implementors would then have the option
+to set this new field instead.
+
+**Note:** If a new model is added to the schema but clients are not updated,
+they'll encounter an object with no information from any known structured model
+when they serialize into their known version of a `NodeResourceSlice`. This
+tells them that they cannot handle the object because the API has been extended.
+
+Drivers can use different structured models by publishing multiple
+`NodeResourceSlice` objects, as long as each model represents a distinct set of
+resources. Whether the information about resources of one particular structured
+model must fit into one NodeResourceSlice object (or be distributed across
+many) depends on how that particular structured model describes its resources. In
+all cases, the size of each object is a hard limit and one must take this into
+account when designing a structured model and preparing NodeResourceSlice objects
+for it.
+
+Below is an example of a driver that provides two discrete GPU cards using the
+`namedResourcesWithAttributes` model described above:
+
+```yaml
+kind: NodeResourceSlice
+apiVersion: resource.k8s.io/v1alpha2
+...
+spec:
+  nodeName: worker-1
+  driverName: cards.dra.example.com
+  namedResourcesWithAttributes:
+  - name: gpu-0
+    attributes:
+    - name: UUID
+      string: GPU-ceea231c-4257-7af7-6726-efcb8fc2ace9
+    - name: driverVersion
+      string: 1.2.3
+    - name: runtimeVersion
+      string: 11.1.42
+    - name: memory
+      quantity: 16Gi
+    - name: productName:
+      string: ACME T1000 32GB
+  - name: gpu-1
+    attributes:
+    - name: UUID
+      string: GPU-6aa0af9e-a2be-88c8-d2b3-2240d25318d7
+    - name: driverVersion
+      string: 1.2.3
+    - name: runtimeVersion
+      string: 11.1.42
+    - name: memory
+      quantity: 32Gi
+    - name: productName:
+      string: ACME A4-PCIE-40GB
+```
+
+Where "gpu-0" represents one type of card and "gpu-1" represents another (with
+the attributes hanging off each serving to "define" their individual
```

Comment:
simpler:
s/attributes hanging off each serving to "define"/related attributes defining/